pkg/gpu/ebpf/c/gpu.c

```diff
@@ -0,0 +1,139 @@
+#define BPF_NO_PRESERVE_ACCESS_INDEX
+#define BPF_NO_GLOBAL_DATA
+
+#include "ktypes.h"
+#include "bpf_metadata.h"
+#include "bpf_helpers.h"
+#include "bpf_helpers_custom.h"
+#include "bpf_tracing.h"
+#include "compiler.h"
+#include "map-defs.h"
+
+#include "types.h"
+
+char __license[] SEC("license") = "GPL";
+
+BPF_PERF_EVENT_ARRAY_MAP(cuda_events, cuda_event_header_t);
+BPF_LRU_MAP(cuda_alloc_cache, __u64, cuda_alloc_request_args_t, 1024)
+
+// cudaLaunchKernel receives the dim3 argument by value, which gets translated as
+// a 64 bit register with the x and y values in the lower and upper 32 bits respectively,
+// and the z value in a separate register. This function decodes those values into a dim3 struct.
+static inline void load_dim3(__u64 xy, __u64 z, dim3 *dst) {
+    __u64 mask = 0xffffffff;
+    dst->x = xy & mask;
+    dst->y = xy >> 32;
+    dst->z = z;
+}
+
+SEC("uprobe/cudaLaunchKernel")
+int BPF_UPROBE(uprobe__cudaLaunchKernel, const void *func, __u64 grid_xy, __u64 grid_z, __u64 block_xy, __u64 block_z, void **args) {
+    cuda_kernel_launch_t launch_data;
+    __u64 shared_mem = 0;
+    __u64 stream = 0;
+
+    shared_mem = PT_REGS_USER_PARM7(ctx);
+    stream = PT_REGS_USER_PARM8(ctx);
+
+    __builtin_memset(&launch_data, 0, sizeof(launch_data));
+
+    load_dim3(grid_xy, grid_z, &launch_data.grid_size);
+    load_dim3(block_xy, block_z, &launch_data.block_size);
+    launch_data.header.pid_tgid = bpf_get_current_pid_tgid();
+    launch_data.header.ktime_ns = bpf_ktime_get_ns();
+    launch_data.header.stream_id = (uint64_t)stream;
+    launch_data.header.type = cuda_kernel_launch;
+    launch_data.kernel_addr = (uint64_t)func;
+    launch_data.shared_mem_size = shared_mem;
+
+    log_debug("cudaLaunchKernel: EMIT[1/2] pid_tgid=%llu, ts=%llu", launch_data.header.pid_tgid, launch_data.header.ktime_ns);
+    log_debug("cudaLaunchKernel: EMIT[2/2] kernel_addr=0x%llx, shared_mem=%llu, stream_id=%llu", launch_data.kernel_addr, launch_data.shared_mem_size, launch_data.header.stream_id);
+
+    bpf_ringbuf_output(&cuda_events, &launch_data, sizeof(launch_data), 0);
+
+    return 0;
+}
+
+SEC("uprobe/cudaMalloc")
+int BPF_UPROBE(uprobe__cudaMalloc, void **devPtr, size_t size) {
+    __u64 pid_tgid = bpf_get_current_pid_tgid();
+    cuda_alloc_request_args_t args = { .devPtr = devPtr, .size = size };
+
+    log_debug("cudaMalloc: pid=%llu, devPtr=%llx, size=%lu", pid_tgid, (__u64)devPtr, size);
```

Comment:
do we want this or is it just for debugging purposes?

Reply:
It's for debugging purposes, but it's only included on the debug BPF programs, so it doesn't affect production runtime.